drivers: serial: nrf: Make pull-ups on UART pins configurable

anangl opened the pull request:
This is a follow-up to commit 3656ba5ae9448f7132fe51dc4a595665e549fb84.

Do not enforce pull-up resistors to be enabled on RXD and CTS pins
in nRF UART drivers, as in certain hardware designs this may be
undesirable or may even make certain hardware not working.
Instead, provide devicetree properties that allow enabling of those
resistors when it is actually needed.

Fixes #36782

anangl committed:
drivers: serial: nrf: Make pull-ups on UART pins configurable

This is a follow-up to commit 3656ba5ae9448f7132fe51dc4a595665e549fb84.

Do not enforce pull-up resistors to be enabled on RXD and CTS pins
in nRF UART drivers, as in certain hardware designs this may be
undesirable or may even make certain hardware not working.
Instead, provide devicetree properties that allow enabling of those
resistors when it is actually needed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull reviewed (approved):
LGTM, only one note: it looks like this changes the previous default (pull-ups ON), is this intended?

mbolivar-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1954,6 +1948,11 @@ static int uarte_nrfx_pm_control(const struct device *dev,
 		    (NRF_UARTE_PSEL_DISCONNECTED))
 
+#define UARTE_PULL(idx, pull_up_prop) \
+	COND_CODE_1(UARTE_HAS_PROP(idx, pull_up_prop), \
```
In general DT_PROP should be used instead of DT_HAS_PROP in the case of boolean properties. It will expand to 0 or 1 depending on if the property is present.

anangl replied:
Changed to `UARTE_PROP`.

nordic-krch commented:
What about adding these new properties to nordic DKs?

anangl committed:
boards: nrf: Enable pull-ups on UART pins in Nordic DKs

This is a follow-up to commit 3656ba5ae9448f7132fe51dc4a595665e549fb84.

Enable pull-up resistors on UART pins RXD and CTS in Nordic DK boards
where it is possible that these pins are not connected to anything on
the board and could generate problems if left floating (on some DKs
also the Interface MCU can be disconnected with the switch "nRF only",
hence pull-ups enabled also on uart0 for them).
Those pull-ups can still be disabled (by deleting the added properties)
if necessary to use a given board in some specific hardware setup.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
> LGTM, only one note: it looks like this changes the previous default (pull-ups ON), is this intended?

@gmarull Yes. The intent is to enable the pull-ups only when it is at least potentially necessary.

> What about adding these new properties to nordic DKs?

@nordic-krch I added a commit that enables pull-ups on pins assigned to UARTs that are not connected to anything on Nordic DKs. I don't think there's a need to enable pull-ups in other cases, in particular for UARTs that are connected to the Interface MCUs. Please have a look.

mniestroj reviewed (approved):
LGTM. Thanks a lot!

mniestroj commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -996,7 +996,9 @@ static int uart_nrfx_init(const struct device *dev)
 	if (RX_PIN_USED) {
-		nrf_gpio_cfg_input(RX_PIN, NRF_GPIO_PIN_PULLUP);
+		nrf_gpio_cfg_input(RX_PIN,
+			HAS_PROP(rx_pull_up) ? NRF_GPIO_PIN_PULLUP
```
I've got a second look and propably this HAS_PROP needs to be replaced by PROP.

anangl replied:
Right. Thanks for catching this. I'll fix it promptly.

nordic-krch commented on boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:
```diff
@@ -166,6 +166,7 @@ arduino_serial: &uart1 {
 	status = "okay";
 	current-speed = <115200>;
 	rx-pin = <33>;
+	rx-pull-up;
```
it should also be added to `uart0` which is connected to segger chip but becomes floating after flipping `nRF only` switch. Original fix was due to increased power consumption after flipping this switch.

anangl replied:
Updated.
